sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -363,6 +364,70 @@ trait Row extends Serializable {
     false
   }
 
+  protected def canEqual(other: Any) =
+    other.isInstanceOf[Row] && !other.isInstanceOf[InternalRow]
```

Comment:
Previously we implemented equals in both internal row(`InternalRow`) and external row(`GenericRow`) with very similar logic, so I moved it to the root class `Row`, but using `canEqual` to prevent compare internal row and external row. (`InternalRow` override `canEqual`)